### PR TITLE
[FLINK-5792][TableAPI&SQL] Improve UDTF to support with parameter con…

### PR DESCRIPTION
**Pull request by sunjincheng121**
…structor

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/api/java/functions/utils/CodingUtils.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            String hv = Integer.toHexString(v);
+            if (hv.length() < 2) {
+                stringBuilder.append(0);
+            }
+            stringBuilder.append(hv);
+        }
+        return stringBuilder.toString();
+    }
+
+    public static byte[] toBytes(String hexString) {
```
Comment (zhoukaibo) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/api/java/functions/utils/CodingUtils.java]: byte[] 和 string 之间的转换看能不能用base64编解码，不需要自己实现
JDK里面提供了sun.misc.BASE64Encoder()和sun.misc.BASE64Decoder
我看flink里面也有对 apache commons-codec 包的依赖，里面有 Base64.encodeBase64 和 Base64.decodeBase64

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/api/java/functions/utils/CodingUtils.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.api.java.functions.utils;
+
+import java.security.MessageDigest;
+
+public class CodingUtils {
```
Comment (zhoukaibo) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/api/java/functions/utils/CodingUtils.java]: CodingUtils 名字可以换个具体点的

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 3]
```diff
@@ -324,4 +325,19 @@ object UserDefinedFunctionUtils {
     candidate == classOf[Time] && (expected == classOf[Int] || expected == classOf[JInt]) ||
     candidate == classOf[Timestamp] && (expected == classOf[Long] || expected == classOf[JLong])
 
+  @throws[Exception]
+  def serializable(function: UserDefinedFunction): String = {
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: -> serialize

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 4]
```diff
@@ -324,4 +325,19 @@ object UserDefinedFunctionUtils {
...
+    val byteArrayOutPut = new ByteArrayOutputStream
+    val objectOutPut = new ObjectOutputStream(byteArrayOutPut)
+    objectOutPut.writeObject(function)
+    objectOutPut.flush()
+    Base64.encodeBase64URLSafeString(byteArrayOutPut.toByteArray)
+  }
+
+  @throws[Exception]
+  def deSerializable(data: String): UserDefinedFunction = {
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: -> deserialize

**Pull request closed without merging**